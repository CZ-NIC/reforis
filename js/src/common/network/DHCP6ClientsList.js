/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import moment from "moment";
import { formFieldsSize } from "foris";

DHCP6ClientsList.propTypes = {
    ipv6clients: PropTypes.arrayOf(PropTypes.object),
};

export default function DHCP6ClientsList({ ipv6clients }) {
    return (
        <div className={formFieldsSize}>
            <h2>{_("IPv6 DHCP Client List")}</h2>
            <p>
                {_(
                    "This list contains all devices that are connected to the network through wired or wireless connections using IPv6."
                )}
            </p>
            {ipv6clients.length === 0 ? (
                <p className="text-muted text-center">
                    {_("No clients found.")}
                </p>
            ) : (
                <div className="table-responsive">
                    <table className="table table-hover">
                        <thead className="thead-light">
                            <tr className="text-left">
                                <th>{_("Hostname")}</th>
                                <th>{_("IPv6 Address")}</th>
                                <th>{_("DUID")}</th>
                                <th className="text-center">{_("Expires")}</th>
                            </tr>
                        </thead>
                        <tbody>
                            {ipv6clients.map((client) => (
                                <DHCP6ClientsListItem
                                    key={client.ipv6}
                                    {...client}
                                />
                            ))}
                        </tbody>
                    </table>
                </div>
            )}
        </div>
    );
}

DHCP6ClientsListItem.propTypes = {
    ipv6: PropTypes.string.isRequired,
    expires: PropTypes.number.isRequired,
    duid: PropTypes.string.isRequired,
    hostname: PropTypes.string.isRequired,
};

function DHCP6ClientsListItem({ ipv6, expires, duid, hostname }) {
    return (
        <tr className="text-left">
            <td>{hostname}</td>
            <td>{ipv6}</td>
            <td>{duid}</td>
            <td className="text-center">
                {moment.unix(expires).format("YYYY-MM-DD HH:mm")}
            </td>
        </tr>
    );
}
