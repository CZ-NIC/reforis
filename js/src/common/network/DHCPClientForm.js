/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

import TextInput from '../bootstrap/TextInput';
import {validateDomain} from '../validations';

const HELP_TEXTS = {
    hostname: _('Hostname which will be provided to DHCP server.'),
};

DHCPClientForm.propTypes = {
    formData: propTypes.shape({
        hostname: propTypes.string,
    }).isRequired,
    formErrors: propTypes.shape({
        hostname: propTypes.string
    }),
    setFormValue: propTypes.func.isRequired,
    updateRule: propTypes.func.isRequired,
};

export default function DHCPClientForm({formData, formErrors, setFormValue, updateRule, ...props}) {
    return <TextInput
        label={_('DHCP hostname')}
        value={formData.hostname || ''}
        error={(formErrors || {}).hostname || null}
        helpText={HELP_TEXTS.hostname}

        onChange={setFormValue(
            value => updateRule({hostname: {$set: value}})
        )}

        {...props}
    />
}

export function validateDHCPForm(formData) {
    const error = {hostname: validateDomain(formData.hostname)};
    return error.hostname ? error : null;
}
