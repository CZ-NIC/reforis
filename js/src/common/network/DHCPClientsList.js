/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import moment from "moment";

DHCPClientsList.propTypes = {
    clients: PropTypes.arrayOf(PropTypes.object),
};

export default function DHCPClientsList({ clients }) {
    return (
            <h2>{_("DHCP Client List")}</h2>
            <p>
                {_(
                    "This list contains all devices that are connected to the network through wired or wireless connections."
                )}
            </p>
            {clients.length === 0 ? (
                <p className="text-muted text-center">
                    {_("No DHCP clients found.")}
                </p>
            ) : (
                <div className="table-responsive">
                    <table className="table table-hover">
                        <thead>
                            <tr className="text-center">
                                <th>{_("Expires")}</th>
                                <th>{_("IP Address")}</th>
                                <th>{_("MAC Address")}</th>
                                <th>{_("Hostname")}</th>
                                <th>{_("Active")}</th>
                            </tr>
                        </thead>
                        <tbody>
                            {clients.map((client) => (
                                <DHCPClientsListItem
                                    key={client.ip}
                                    {...client}
                                />
                            ))}
                        </tbody>
                    </table>
                </div>
            )}
        </>
    );
}

DHCPClientsListItem.propTypes = {
    ip: PropTypes.string.isRequired,
    expires: PropTypes.number.isRequired,
    mac: PropTypes.string.isRequired,
    hostname: PropTypes.string.isRequired,
    active: PropTypes.bool.isRequired,
};

function DHCPClientsListItem({ ip, expires, mac, hostname, active }) {
    return (
        <tr className="text-center">
            <td>{moment.unix(expires).format("YYYY-MM-DD HH:mm")}</td>
            <td>{ip}</td>
            <td>{mac}</td>
            <td>{hostname}</td>
            <td>
                <i className={`fas ${active ? "fa-check" : "fa-times"}`} />
            </td>
        </tr>
    );
}
