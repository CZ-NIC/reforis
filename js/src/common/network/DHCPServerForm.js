/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

import NumberInput from 'common/bootstrap/NumberInput';

export const HELP_TEXT = _('Enable this option to automatically assign IP addresses to the devices connected to the router.');

DHCPServerForm.propTypes = {
    formData: propTypes.shape({
        start: propTypes.number,
        limit: propTypes.number,
        lease_time: propTypes.number,
    }).isRequired,
    setFormValue: propTypes.func.isRequired,
    updateRule: propTypes.func.isRequired,
};

export default function DHCPServerForm({formData, updateRule, setFormValue, ...props}) {
    return <>
        <NumberInput
            label={_('DHCP start')}
            value={formData.start}
            min='1'
            required

            onChange={setFormValue(
                value => updateRule({start: {$set: value}})
            )}

            {...props}
        />
        <NumberInput
            label={_('DHCP max leases')}
            value={formData.limit}
            min='1'
            required

            onChange={setFormValue(
                value => updateRule({limit: {$set: value}})
            )}

            {...props}
        />
        <NumberInput
            label={_('Lease time (hours)')}
            value={formData.lease_time}
            min='120'
            required

            onChange={setFormValue(
                value => updateRule({lease_time: {$set: value}})
            )}

            {...props}
        />
    </>
}
