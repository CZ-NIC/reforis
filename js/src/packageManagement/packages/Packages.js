/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";
import {
    Alert, ForisURLs, ForisForm, useAPIGet, API_STATE, Spinner,
} from "foris";

import API_URLs from "common/API";
import LanguageForm from "./LanguageForm";
import PackagesForm from "./PackagesForm";

export default function Packages() {
    return (
        <>
            <h1>{_("Packages")}</h1>
            <p>
                {_(`
Apart from the standard installation, you can optionally select bundles of additional software that'd be installed on 
the router. This software can be selected from the following list. Please note that only software that is part of 
TurrisOS or that has been installed from a package list is maintained by automatic updates system. Software that has 
been installed manually or using opkg is not affected.
        `)}
            </p>
            <ForisForm
                forisConfig={{ endpoint: API_URLs.packages }}
                prepData={prepData}
                prepDataToSubmit={prepDataToSubmit}
                validator={validator}
            >
                <DisableIfUpdaterIsDisabled>
                    <PackagesForm />
                    <LanguageForm />
                </DisableIfUpdaterIsDisabled>
            </ForisForm>
        </>
    );
}

DisableIfUpdaterIsDisabled.propTypes = {
    formData: PropTypes.shape({ enabled: PropTypes.bool }),
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};

export function DisableIfUpdaterIsDisabled({ children, formData, ...props }) {
    const [getEnabledState, getEnabled] = useAPIGet(API_URLs.updatesEnabled);

    useEffect(() => {
        getEnabled();
    }, [getEnabled]);

    if (getEnabledState.state !== API_STATE.SUCCESS) {
        return <Spinner />;
    }

    const isDisabled = !getEnabledState.data.enabled;

    const childrenWithFormProps = React.Children.map(
        children,
        (child) => React.cloneElement(child, {
            ...props,
            formData,
            disabled: isDisabled,
        }),
    );

    if (!isDisabled) return childrenWithFormProps;

    return (
        <>
            <DisabledUpdaterAlert />
            <div className="text-muted">
                {childrenWithFormProps}
            </div>
        </>
    );
}

function DisabledUpdaterAlert() {
    const message = _(
        `Please enable <a href="${ForisURLs.packageManagement.updateSettings}">automatic updates</a> to manage packages and languages.`,
    );
    // <Alert /> is used instead of context because warning isn't result of any action on this page
    return (
        <Alert type="warning">
            <span dangerouslySetInnerHTML={{ __html: message }} />
        </Alert>
    );
}

function prepData(formData) {
    formData.package_lists = formData.package_lists
        .filter((_package) => !_package.hidden);
    return formData;
}

function prepDataToSubmit(formData) {
    const packages = formData.package_lists
        .filter((_package) => _package.enabled)
        .map((_package) => {
            const options = _package.options.map((option) => ({
                name: option.name,
                enabled: option.enabled,
            }));
            return {
                name: _package.name,
                options,
            };
        });

    const languages = formData.languages
        .filter((language) => language.enabled)
        .map((language) => language.code);

    return {
        languages,
        package_lists: packages,
    };
}

// Hack to disable submit button
function validator(formData) {
    if (!formData.enabled) return { enabled: true };
    return null;
}
