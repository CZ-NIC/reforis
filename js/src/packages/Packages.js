/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import PropTypes from 'prop-types';

import API_URLs from 'common/API';
import Alert from 'common/bootstrap/Alert';
import {ForisURLs} from 'common/constants';
import ForisForm from 'form/ForisForm';

import LanguageForm from './LanguageForm';
import PackagesForm from './PackagesForm';

export default function Packages() {
    return <>
        <h1>{_('Packages')}</h1>
        <p>{_(`
Apart from the standard installation, you can optionally select bundles of additional software that'd be installed on 
the router. This software can be selected from the following list. Please note that only software that is part of 
TurrisOS or that has been installed from a package list is maintained by automatic updates system. Software that has 
been installed manually or using opkg is not affected.
        `)}</p>
        <ForisForm
            forisConfig={{endpoint: API_URLs.packages}}
            prepData={prepData}
            prepDataToSubmit={prepDataToSubmit}
            validator={validator}
        >
            <DisableIfUpdaterIsDisabled>
                <PackagesForm/>
                <LanguageForm/>
            </DisableIfUpdaterIsDisabled>
        </ForisForm>
    </>
}

DisableIfUpdaterIsDisabled.propTypes = {
    formData: PropTypes.shape({enabled: PropTypes.bool.isRequired}),
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ])
};

export function DisableIfUpdaterIsDisabled({children, ...props}) {
    const isDisabled = !props.formData.enabled;
    const childrenWithFormProps =
        React.Children.map(children, child =>
            React.cloneElement(child, {
                ...props,
                disabled: isDisabled,
            })
        );

    if (!isDisabled)
        return childrenWithFormProps;

    return <>
        <DisabledUpdaterAlert/>
        <div className='text-muted'>
            {childrenWithFormProps}
        </div>
    </>
}

DisabledUpdaterAlert.propTypes = {
    formData: PropTypes.shape({enabled: PropTypes.bool.isRequired})
};

function DisabledUpdaterAlert() {
    const message = _(
        `Please enable <a href="${ForisURLs.updates}">automatic updates</a> to manage packages and languages.`
    );
    return <Alert type='warning'>
        <span dangerouslySetInnerHTML={{__html: message}}/>
    </Alert>
}

function prepData(formData) {
    formData.user_lists = formData.user_lists
        .filter(_package => !_package.hidden);
    return formData
}

function prepDataToSubmit(formData) {
    const packages = formData.user_lists
        .filter(_package => _package.enabled)
        .map(_package => _package.name);

    const languages = formData.languages
        .filter(language => language.enabled)
        .map(language => language.code);

    return {languages: languages, user_lists: packages}
}

// Hack to disable submit button
function validator(formData) {
    if (!formData.enabled)
        return {enabled: true};
    return null;
}
