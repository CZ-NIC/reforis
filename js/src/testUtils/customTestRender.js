/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
