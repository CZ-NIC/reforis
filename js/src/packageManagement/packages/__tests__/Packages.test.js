/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import diffSnapshot from "snapshot-diff";

import { cleanup, render, wait } from "foris/testUtils/customTestRender";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from "jest-mock-axios";

import packagesFixture from "./__fixtures__/packages";
import Packages from "../Packages";

describe("<Packages/>", () => {
    let firstRender;
    let queryByText;

    beforeEach(async () => {
        const renderRes = render(<Packages/>);
        queryByText = renderRes.queryByText;
        await wait(() => expect(mockAxios.get).toBeCalledWith("/reforis/api/packages", expect.anything()));
        mockAxios.mockResponse({data: packagesFixture(true)});
        await wait(() => renderRes.getByText("Enabled package title"));
        firstRender = renderRes.asFragment();
    });

    it("should handle error", async () => {
        const { getByText } = render(<Packages/>);
        mockJSONError();
        await wait(() => {
            expect(getByText("An error occurred while fetching data.")).toBeTruthy();
        });
    });

    it("Updates enabled", () => {
        expect(firstRender).toMatchSnapshot()
    });

    it("Updates disabled", async () => {
        cleanup();
        const {getByText, asFragment} = render(<Packages/>);
        await wait(() => expect(mockAxios.get).toBeCalledWith("/reforis/api/packages", expect.anything()));
        mockAxios.mockResponse({data: packagesFixture(false)});

        await wait(() => getByText("Enabled package title"));

        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it("Test hidden.", () => {
        const HTMLHiddenPackageMessage = queryByText("Hidden package msg");
        expect(HTMLHiddenPackageMessage).toBeNull();
    })
});
