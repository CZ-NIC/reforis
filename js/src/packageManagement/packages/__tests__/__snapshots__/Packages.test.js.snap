// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Packages/> Should render: updates disabled. 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -12,10 +12,26 @@
    </p>
    <div
      class=\\"card p-4 col-sm-12 col-lg-12 p-0 mb-3\\"
    >
      <form>
+       <div
+         class=\\"alert  alert-warning\\"
+       >
+         <span>
+           Please enable 
+           <a
+             href=\\"/reforis/package-management/update-settings\\"
+           >
+             automatic updates
+           </a>
+            to manage packages and languages.
+         </span>
+       </div>
+       <div
+         class=\\"text-muted\\"
+       >
          <h2>
            Packages List
          </h2>
          <div
            class=\\"container-fluid\\"
@@ -33,10 +49,11 @@
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      checked=\\"\\"
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"1\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -61,10 +78,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"2\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -89,10 +107,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"3\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -117,10 +136,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"4\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -145,10 +165,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"5\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -173,10 +194,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"6\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -201,33 +223,34 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"7\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
                      for=\\"7\\"
                    >
                      LXC utilities
                    </label>
                    <span
-                   class=\\"badge badge-primary\\"
+                     class=\\"badge badge-primary badge-disabled\\"
                      title=\\"Using external storage is highly suggested for this package list usage. Software it contains can wear out flash storages as it performs increased amount of writes.\\"
                    >
                      External storage
                    </span>
                    <span
-                   class=\\"badge badge-info\\"
+                     class=\\"badge badge-info badge-disabled\\"
                      title=\\"Software in this package list consumes possibly higher amount of RAM to run. It is not suggested to use it with small memory.\\"
                    >
                      High memory usage
                    </span>
                    <span
-                   class=\\"badge badge-secondary\\"
+                     class=\\"badge badge-secondary badge-disabled\\"
                      title=\\"This functionality is usable only for advanced users.\\"
                    >
                      Advanced users
                    </span>
                    <small
@@ -247,10 +270,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"8\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -412,10 +436,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"14\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -440,10 +465,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"15\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -468,10 +494,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"16\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -496,10 +523,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"17\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -524,10 +552,11 @@
                  <div
                    class=\\"custom-control custom-checkbox \\"
                  >
                    <input
                      class=\\"custom-control-input\\"
+                     disabled=\\"\\"
                      id=\\"18\\"
                      type=\\"checkbox\\"
                    />
                    <label
                      class=\\"custom-control-label\\"
@@ -543,15 +572,17 @@
                  </div>
                </div>
              </div>
            </div>
          </div>
+       </div>
        <div
          class=\\"text-right\\"
        >
          <button
            class=\\"btn btn-primary  col-sm-12 col-md-3 col-lg-2\\"
+           disabled=\\"\\"
            type=\\"submit\\"
          >
            Save
          </button>
        </div>"
`;

exports[`<Packages/> Should render: updates enabled. 1`] = `
<DocumentFragment>
  <h1>
    Packages
  </h1>
  <p>
    
Apart from the standard installation, you can optionally select bundles of additional software that'd be installed on 
the router. This software can be selected from the following list. Please note that only software that is part of 
TurrisOS or that has been installed from a package list and not marked as community-supported is actually supported
by Turris team.
        
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-3"
  >
    <form>
      <h2>
        Packages List
      </h2>
      <div
        class="container-fluid"
      >
        <div
          class="packages-list"
        >
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  checked=""
                  class="custom-control-input"
                  id="1"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="1"
                >
                  Enabled package title
                </label>
                <small
                  class="form-text text-muted"
                >
                  Enabled package description
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="2"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="2"
                >
                  Extensions of network protocols for 3G/LTE
                </label>
                <small
                  class="form-text text-muted"
                >
                  Support for additional protocols and connection types.
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="3"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="3"
                >
                  Device detection
                </label>
                <small
                  class="form-text text-muted"
                >
                  Software for detecting new devices on local network (EXPERIMENTAL).
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="4"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="4"
                >
                  DVB tuner
                </label>
                <small
                  class="form-text text-muted"
                >
                  Software for sharing television received by a DVB tuner on Turris. Does not include device drivers.
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="5"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="5"
                >
                  SSH Honeypot
                </label>
                <small
                  class="form-text text-muted"
                >
                  Trap for password-guessing robots on SSH.
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="6"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="6"
                >
                  LuCI extensions
                </label>
                <small
                  class="form-text text-muted"
                >
                  Several addional tabs and controls for the advanced LuCI interface.
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="7"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="7"
                >
                  LXC utilities
                </label>
                <span
                  class="badge badge-primary"
                  title="Using external storage is highly suggested for this package list usage. Software it contains can wear out flash storages as it performs increased amount of writes."
                >
                  External storage
                </span>
                <span
                  class="badge badge-info"
                  title="Software in this package list consumes possibly higher amount of RAM to run. It is not suggested to use it with small memory."
                >
                  High memory usage
                </span>
                <span
                  class="badge badge-secondary"
                  title="This functionality is usable only for advanced users."
                >
                  Advanced users
                </span>
                <small
                  class="form-text text-muted"
                >
                  Set of utilities to manage Linux Containers (lightweight virtualization technology).
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="8"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="8"
                >
                  NAS
                </label>
                <small
                  class="form-text text-muted"
                >
                  Services allowing to connect a disk to the router and use it as network data store.
                </small>
              </div>
            </div>
            <div
              class="option"
            >
              <div
                class="form-group"
              >
                <div
                  class="custom-control custom-checkbox "
                >
                  <input
                    checked=""
                    class="custom-control-input"
                    disabled=""
                    id="9"
                    type="checkbox"
                  />
                  <label
                    class="custom-control-label"
                    for="9"
                  >
                    Samba
                  </label>
                  <small
                    class="form-text text-muted"
                  >
                    Implementation of SMB network protocol.
                  </small>
                </div>
              </div>
              <div
                class="form-group"
              >
                <div
                  class="custom-control custom-checkbox "
                >
                  <input
                    class="custom-control-input"
                    disabled=""
                    id="10"
                    type="checkbox"
                  />
                  <label
                    class="custom-control-label"
                    for="10"
                  >
                    DLNA
                  </label>
                  <small
                    class="form-text text-muted"
                  >
                    Digital media sharing server.
                  </small>
                </div>
              </div>
              <div
                class="form-group"
              >
                <div
                  class="custom-control custom-checkbox "
                >
                  <input
                    checked=""
                    class="custom-control-input"
                    disabled=""
                    id="11"
                    type="checkbox"
                  />
                  <label
                    class="custom-control-label"
                    for="11"
                  >
                    Transmission
                  </label>
                  <small
                    class="form-text text-muted"
                  >
                    BitTorrent client.
                  </small>
                </div>
              </div>
              <div
                class="form-group"
              >
                <div
                  class="custom-control custom-checkbox "
                >
                  <input
                    class="custom-control-input"
                    disabled=""
                    id="12"
                    type="checkbox"
                  />
                  <label
                    class="custom-control-label"
                    for="12"
                  >
                    mdadm RAID
                  </label>
                  <span
                    class="badge badge-secondary badge-disabled"
                    title="This functionality is usable only for advanced users."
                  >
                    Advanced users
                  </span>
                  <small
                    class="form-text text-muted"
                  >
                    Software RAID storage support using mdadm.
                  </small>
                </div>
              </div>
              <div
                class="form-group"
              >
                <div
                  class="custom-control custom-checkbox "
                >
                  <input
                    class="custom-control-input"
                    disabled=""
                    id="13"
                    type="checkbox"
                  />
                  <label
                    class="custom-control-label"
                    for="13"
                  >
                    Encrypted Storage
                  </label>
                  <small
                    class="form-text text-muted"
                  >
                    Add support to access encrypted storage devices using dm-crypt.
                  </small>
                </div>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="14"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="14"
                >
                  Internet connection speed measurement
                </label>
                <small
                  class="form-text text-muted"
                >
                  Actively measures speed of Internet connection using netmetr.cz service.
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="15"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="15"
                >
                  OpenVPN
                </label>
                <small
                  class="form-text text-muted"
                >
                  An easy setup of OpenVPN server from Foris.
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="16"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="16"
                >
                  Pakon
                </label>
                <small
                  class="form-text text-muted"
                >
                  Software for in depth monitoring of your traffic.
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="17"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="17"
                >
                  Print server
                </label>
                <small
                  class="form-text text-muted"
                >
                  Services allowing to connect a printer to the router and use it for remote printing.
                </small>
              </div>
            </div>
          </div>
          <div
            class="package"
          >
            <div
              class="form-group"
            >
              <div
                class="custom-control custom-checkbox "
              >
                <input
                  class="custom-control-input"
                  id="18"
                  type="checkbox"
                />
                <label
                  class="custom-control-label"
                  for="18"
                >
                  Tor
                </label>
                <small
                  class="form-text text-muted"
                >
                  Service to increase anonymity on the Internet.
                </small>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-md-3 col-lg-2"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
